From 6d7989c4a48c73120137d72a8135c2de3a1078b3 Mon Sep 17 00:00:00 2001
From: John Benediktsson <mrjbq7@gmail.com>
Date: Thu, 18 Sep 2025 22:02:22 -0700
Subject: [PATCH 104/133] std.fmt: migrate bufPrintZ to bufPrintSentinel
 (#25260)
X-Developer-Signature: v=1; a=openpgp-sha256; l=5674; i=xtex@astrafall.org;
 h=from:subject; bh=ipXTHaEzdX4ii0O5AU2TotxYK5xg1bIyjWXEMwmoGRM=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn78rhbnOZmNNqZesf/23tfe4ELy6Rz9gYdy3m+HU1ua
 PwewFLVMZGFQYyLwVJMkaXIsMGbVSedX3RZuSzMHFYmkCHSIg0MDAwMLAx8uYl5pUY6Rnqm2oZ6
 hoY6RjpGDFycAjDVlt8YfrMm3XPj2F3SO++v/+ygXdF8qW8SNkm91LmUwsOdUzL/eAQjw443uQd
 cfgcstnn+fH2qxmNt/TUS51WYK6N1WE3VOlxy2QA=
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

(cherry picked from commit 37ecaae6390d238dce21ef4b97e6994dd229c2c3)
---
 lib/std/fmt.zig           | 14 ++++++++++++--
 lib/std/fs.zig            |  3 ++-
 lib/std/meta.zig          |  2 +-
 lib/std/os.zig            |  4 ++--
 lib/std/posix.zig         |  2 +-
 lib/std/process/Child.zig |  3 ++-
 6 files changed, 20 insertions(+), 8 deletions(-)

diff --git a/lib/std/fmt.zig b/lib/std/fmt.zig
index c32d73845562..629a6529702f 100644
--- a/lib/std/fmt.zig
+++ b/lib/std/fmt.zig
@@ -617,9 +617,19 @@ pub fn bufPrint(buf: []u8, comptime fmt: []const u8, args: anytype) BufPrintErro
     return w.buffered();
 }
 
+/// Deprecated in favor of `bufPrintSentinel`
 pub fn bufPrintZ(buf: []u8, comptime fmt: []const u8, args: anytype) BufPrintError![:0]u8 {
-    const result = try bufPrint(buf, fmt ++ "\x00", args);
-    return result[0 .. result.len - 1 :0];
+    return try bufPrintSentinel(buf, fmt, args, 0);
+}
+
+pub fn bufPrintSentinel(
+    buf: []u8,
+    comptime fmt: []const u8,
+    args: anytype,
+    comptime sentinel: u8,
+) BufPrintError![:sentinel]u8 {
+    const result = try bufPrint(buf, fmt ++ [_]u8{sentinel}, args);
+    return result[0 .. result.len - 1 :sentinel];
 }
 
 /// Count the characters needed for format.
diff --git a/lib/std/fs.zig b/lib/std/fs.zig
index ed595b94fc96..266be39c92c8 100644
--- a/lib/std/fs.zig
+++ b/lib/std/fs.zig
@@ -616,9 +616,10 @@ pub fn selfExePath(out_buffer: []u8) SelfExePathError![]u8 {
                 var path_it = mem.tokenizeScalar(u8, PATH, path.delimiter);
                 while (path_it.next()) |a_path| {
                     var resolved_path_buf: [max_path_bytes - 1:0]u8 = undefined;
-                    const resolved_path = std.fmt.bufPrintZ(&resolved_path_buf, "{s}/{s}", .{
+                    const resolved_path = std.fmt.bufPrintSentinel(&resolved_path_buf, "{s}/{s}", .{
                         a_path,
                         std.os.argv[0],
+                        0,
                     }) catch continue;
 
                     var real_path_buf: [max_path_bytes]u8 = undefined;
diff --git a/lib/std/meta.zig b/lib/std/meta.zig
index 019b0e3910c5..7b12094d8a51 100644
--- a/lib/std/meta.zig
+++ b/lib/std/meta.zig
@@ -935,7 +935,7 @@ fn CreateUniqueTuple(comptime N: comptime_int, comptime types: [N]type) type {
         @setEvalBranchQuota(10_000);
         var num_buf: [128]u8 = undefined;
         tuple_fields[i] = .{
-            .name = std.fmt.bufPrintZ(&num_buf, "{d}", .{i}) catch unreachable,
+            .name = std.fmt.bufPrintSentinel(&num_buf, "{d}", .{i}, 0) catch unreachable,
             .type = T,
             .default_value_ptr = null,
             .is_comptime = false,
diff --git a/lib/std/os.zig b/lib/std/os.zig
index b067b93b8a06..f2519a62e902 100644
--- a/lib/std/os.zig
+++ b/lib/std/os.zig
@@ -132,7 +132,7 @@ pub fn getFdPath(fd: std.posix.fd_t, out_buffer: *[max_path_bytes]u8) std.posix.
         },
         .linux, .serenity => {
             var procfs_buf: ["/proc/self/fd/-2147483648\x00".len]u8 = undefined;
-            const proc_path = std.fmt.bufPrintZ(procfs_buf[0..], "/proc/self/fd/{d}", .{fd}) catch unreachable;
+            const proc_path = std.fmt.bufPrintSentinel(procfs_buf[0..], "/proc/self/fd/{d}", .{fd}, 0) catch unreachable;
 
             const target = posix.readlinkZ(proc_path, out_buffer) catch |err| {
                 switch (err) {
@@ -149,7 +149,7 @@ pub fn getFdPath(fd: std.posix.fd_t, out_buffer: *[max_path_bytes]u8) std.posix.
         },
         .solaris, .illumos => {
             var procfs_buf: ["/proc/self/path/-2147483648\x00".len]u8 = undefined;
-            const proc_path = std.fmt.bufPrintZ(procfs_buf[0..], "/proc/self/path/{d}", .{fd}) catch unreachable;
+            const proc_path = std.fmt.bufPrintSentinel(procfs_buf[0..], "/proc/self/path/{d}", .{fd}, 0) catch unreachable;
 
             const target = posix.readlinkZ(proc_path, out_buffer) catch |err| switch (err) {
                 error.UnsupportedReparsePointType => unreachable,
diff --git a/lib/std/posix.zig b/lib/std/posix.zig
index a5c09c5cf2b9..3d6cde066776 100644
--- a/lib/std/posix.zig
+++ b/lib/std/posix.zig
@@ -488,7 +488,7 @@ fn fchmodat2(dirfd: fd_t, path: []const u8, mode: mode_t, flags: u32) FChmodAtEr
         return error.OperationNotSupported;
 
     var procfs_buf: ["/proc/self/fd/-2147483648\x00".len]u8 = undefined;
-    const proc_path = std.fmt.bufPrintZ(procfs_buf[0..], "/proc/self/fd/{d}", .{pathfd}) catch unreachable;
+    const proc_path = std.fmt.bufPrintSentinel(procfs_buf[0..], "/proc/self/fd/{d}", .{pathfd}, 0) catch unreachable;
     while (true) {
         const res = system.chmod(proc_path, mode);
         switch (errno(res)) {
diff --git a/lib/std/process/Child.zig b/lib/std/process/Child.zig
index ca600b5b7f9f..50157d52d956 100644
--- a/lib/std/process/Child.zig
+++ b/lib/std/process/Child.zig
@@ -1324,10 +1324,11 @@ fn windowsMakeAsyncPipe(rd: *?windows.HANDLE, wr: *?windows.HANDLE, sattr: *cons
     const pipe_path = blk: {
         var tmp_buf: [128]u8 = undefined;
         // Forge a random path for the pipe.
-        const pipe_path = std.fmt.bufPrintZ(
+        const pipe_path = std.fmt.bufPrintSentinel(
             &tmp_buf,
             "\\\\.\\pipe\\zig-childprocess-{d}-{d}",
             .{ windows.GetCurrentProcessId(), pipe_name_counter.fetchAdd(1, .monotonic) },
+            0,
         ) catch unreachable;
         const len = std.unicode.wtf8ToWtf16Le(&tmp_bufw, pipe_path) catch unreachable;
         tmp_bufw[len] = 0;
-- 
2.52.0

